src/main/java/uk/ac/ebi/eva/evaseqcol/refget/SHA512Calculator.java

```diff
@@ -1,16 +1,29 @@
 package uk.ac.ebi.eva.evaseqcol.refget;
 
 import javax.xml.bind.DatatypeConverter;
+import java.nio.charset.StandardCharsets;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Base64;
 
 public class SHA512Calculator extends ChecksumCalculator{
 
     @Override
+    /**
+     * Calculate the refget checksum (ga4gh) of the given sequence
+     * */
     public String calculateChecksum(String sequence) {
         String sequence1 = sequence.replaceAll("[\\W]|_", ""); // Remove non-alphanumeric characters
-        String sequence2 =  sequence1.toUpperCase();
-        return SHA512Hash(sequence2);
+        String sequence2 =  SHA512Hash(sequence1.toUpperCase());
+        byte[] sequenceBytes = sequence2.getBytes();
+
+        // Encode the first 24 bytes with Base64
+        byte[] truncatedSequence = Arrays.copyOfRange(sequenceBytes, 0, 24);
+        String encodedBase64Sequence = Base64.getUrlEncoder().encodeToString(truncatedSequence);
+
+        String finalSequence = "SQ." + encodedBase64Sequence;
+        return finalSequence;
     }
 
     public String SHA512Hash(String text) {
```

Comment:
I would suggest to stick to bytes array for the return rather than convert back to text. That's most likely the source of the discrepancy you're seeing.

Reply:
@tcezard 
I don't get it.
The final sequence isn't supposed to be a string ?
Like this : `"SQ.lZyxiD_ByprhOUzrR1o1bq0ezO_1gkrn"`

Reply:
Another remark regarding the **python implementation** of the `trunc512_digest()` ([here](https://github.com/waterflow80/seqcol/blob/master/seqcol/hash_functions.py)).
When trying to calculate the digest of a sequence (`BK006935.2`), I got this output :
`u'93cd477947888f57045a78a086c00c56065fa98e36974ff9` 
which I didn't quite understand it actually. Because I thought I was supposed to return:
`SQ.lZyxiD_ByprhOUzrR1o1bq0ezO_1gkrn`

Reply:
Sorry I should have made it clearer: 
In `SHA512Hash`  you're converting the byte array to Hex using `printHexBinary`.
You don't need to do that because you are subsequently converting back to byte array in `byte[] sequenceBytes = sequence2.getBytes();`
Also the conversion to Hex string is the reason you don't get the correct checksum at the end

the correct conversion should be:
text -> byte array -> digest -> truncate to 24 bytes -> base64 url safe encoding -> text
You are doing 
text -> byte array -> digest -> Hexstring -> text -> byte array -> truncate to 24 bytes -> base64 url safe encoding -> text
I hope this makes more sense

Reply:
Here is a version that works in my test:
```java
    public String calculateChecksum(String sequence) {
        String sequence1 = sequence.replaceAll("[\\W]|_", ""); // Remove non-alphanumeric characters
        byte[] digest =  SHA512Hash(sequence1.toUpperCase());

        // Encode the first 24 bytes with Base64
        byte[] truncatedSequence = Arrays.copyOfRange(digest, 0, 24);
        String encodedBase64Sequence = Base64.getUrlEncoder().encodeToString(truncatedSequence);
        return "SQ." + encodedBase64Sequence;
    }

    public byte[] SHA512Hash(String text) {
        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        md.update(text.getBytes());
        return md.digest();
    }
```

Reply:
@tcezard 
Perfectly clear.
Thanks Tim.
It works fine now.

Reply:
I fixed it [here](https://github.com/EBIvariation/eva-seqcol/pull/3/commits/66b8d3cb62966019feb890347214c23bea5f1536).